### PR TITLE
Add input should-create-pr

### PR DESCRIPTION
**Pull request by jackkoenig**
This optional input takes a script as an argument that lets the user decide if the PR should be created based on the candidate version of firtool.

Tested a bit here: https://github.com/jackkoenig/chisel/actions/workflows/cd-circt.yml

Showing the 3 possibilities:
* Success (script returns 0)  - https://github.com/jackkoenig/chisel/actions/runs/10257662561 (note the action failed because I don't have the token set up to open a PR)
* Skip (script returns 1) - https://github.com/jackkoenig/chisel/actions/runs/10257753361
* Abort (script returns not 0 nor 1) - https://github.com/jackkoenig/chisel/actions/runs/10257762647

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -38,6 +38,15 @@ inputs:
...
     default: './etc/circt.json'
+  should-create-pr:
+    description: |
+      An optional script to decide whether to open the Pull Request or not.
+      If there is a new version of CIRCT, the action will run the provided
+      script. A return value of 0 means "open the PR". A return value of 1
+      means "do not open the PR". Any other return value will cause the action
+      to exit with an error.
+    type: string
+    default: ''
```
Comment (seldridge) [inline action.yml]: Is there a fixed default value that we can use that does the right thing and is also self-documenting?

```suggestion
    default: 'true'
```

File: README.md [hunk 2]
```diff
@@ -36,6 +36,15 @@ binaries/libraries in a downstream project.
...
     circt-config: ''
 
+    # An optional script to decide whether to open the Pull Request or not.
+    # If there is a new version of CIRCT, the action will run the provided
+    # script. A return value of 0 means "open the PR". A return value of 1
+    # means "do not open the PR". Any other return value will cause the action
+    # to exit with an error.
+    #
+    # Default: ''
+    should-create-pr: ''
```
Comment (seldridge) [inline README.md:48]: Thanks for keeping this up-to-date. It is unfortunate that this isn't auto-generated in some way. I expect YAML -> JSON -> jq could do this. Out-of-scope.

File: action.yml [hunk 3]
```diff
@@ -86,6 +95,33 @@ runs:
         VERSION=$(git tag --list 'firtool-*' | tail -n1)
         echo version=$VERSION >> $GITHUB_OUTPUT
         echo Latest version is: \`$VERSION\` >> $GITHUB_STEP_SUMMARY
+    - name: Decide whether or not to open PR
+      id: open-pr
+      shell: bash
+      run: |
+        VERSION=${{ steps.get-version-latest.outputs.version }}
+        if [[ ${{ steps.get-version-current.outputs.version }} != "$VERSION" ]]; then
```
Comment (seldridge) [inline action.yml]: Can the condition be inverted and changed to an early exit?

File: action.yml [hunk 4]
```diff
@@ -86,6 +95,33 @@ runs:
...
+            else
+              RESULT=$?
+            fi
+          else
+            RESULT="0"
+          fi
+          if [[ "$RESULT" -eq "0" ]]; then
+            echo "result=yes" >> $GITHUB_OUTPUT
+          elif [[ "$RESULT" -eq "1" ]]; then
+            echo "::notice::User provided ${{ inputs.should-create-pr }} returned 1, skipping PR creation..."
```
Comment (seldridge) [inline action.yml]: What does `::notice::User` do?

File: action.yml [hunk 5]
```diff
@@ -86,6 +95,33 @@ runs:
...
+            if $(downstream/${{ inputs.should-create-pr }} "$VERSION"); then
+              RESULT="0"
+            else
+              RESULT=$?
+            fi
+          else
+            RESULT="0"
+          fi
+          if [[ "$RESULT" -eq "0" ]]; then
+            echo "result=yes" >> $GITHUB_OUTPUT
```
Comment (seldridge) [inline action.yml]: Can the PR use a GitHub-documented boolean instead of a `yes`/`no`. No Javascript shenanigans here...

File: action.yml [hunk 6]
```diff
@@ -38,6 +38,15 @@ inputs:
       that this is a JSON object containing a key "version".
     type: string
     default: './etc/circt.json'
+  should-create-pr:
+    description: |
+      An optional script to decide whether to open the Pull Request or not.
+      If there is a new version of CIRCT, the action will run the provided
+      script. A return value of 0 means "open the PR". A return value of 1
+      means "do not open the PR". Any other return value will cause the action
+      to exit with an error.
```
Comment (seldridge) [inline action.yml]: Documentation needs to indicate that this takes a single positional argument.

**Review by seldridge — APPROVED** [review]
This is a great addition. Comments throughout. Land when you want.

This will need a minor version bump and a forcible update of the major tag. (Yes, this is how you are supposed to do it...) <- Notes for myself. I will take care of this.

File: action.yml (hunk 4 above)
Comment (jackkoenig) [reply action.yml]: Documented-ish here: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-notice-message

Here's what it looks like in the Summary page:
Screenshot 2024-08-06 at 8 36 26 AM

And in the log of the job:
Screenshot 2024-08-06 at 8 37 13 AM

Kinda nice IMO!

Comment (seldridge) [reply action.yml]: Ah, cool. 👍

**Review by seldridge — APPROVED** [review]
LGTM

**Pull request merged**